File: lintcode/python/0153_combination_sum_II.py
#!/usr/bin/python -t

# dfs

class Solution:
    """
    @param num: Given the candidate numbers
    @param target: Given the target number
    @return: All the combinations that sum to target
    """
    def combinationSum2(self, num, target):
        # write your code here
        num.sort()
        
        ret = []
        
        self.dfs(num, target, 0, [], ret)
        
        return ret
        
    def dfs(self, num, target, start, com, ret):
        if target < 0:
            return
        
        if target == 0:
            ret.append(list(com))
            
        for i in range(start, len(num)):
            if i != start and num[i] == num[i-1]:
                continue
            
            if target < num[i]:
                break
            com.append(num[i])
            self.dfs(num, target-num[i], i+1, com, ret)
            com.pop()
            
        return
    
    

File: lintcode/golang/0153_combination_sum_II.go
/**
 * @param num: Given the candidate numbers
 * @param target: Given the target number
 * @return: All the combinations that sum to target
 */
 
import "sort"

func combinationSum2 (num []int, target int) [][]int {
    // write your code here
    if len(num) == 0 {
        return [][]int{}
    }
    
    sort.Ints(num)
    ret := [][]int{}
    
    var dfs func ([]int, int, int, []int, *[][]int)
    
    dfs = func (num []int, target int, index int, path []int, ret *[][]int) {
        if target == 0 {
            tmp := make([]int, len(path))
            copy(tmp, path)
            *ret = append(*ret, tmp)
            return
        }
        
        for i := index; i < len(num); i++ {
            if i != index && num[i] == num[i-1] {
                continue
            }
            if target - num[i] < 0 {
                break
            }
            path = append(path, num[i])
            dfs(num, target-num[i], i+1, path, ret)
            path = path[:len(path)-1]
        }
        
        return
    }
    
    dfs(num, target, 0, []int{}, &ret)
    
    return ret
}

